### ISSUE TITLE
[analysis] Fix :protocol-impls implementation

### ISSUE THREAD

**Issue by iarenaza** [clj-kondo/clj-kondo#1651]
**Version**

Current version from master branch in Github.

**Platform**

The bug is present in both the JVM version and the native version.

**Editor**

Vanilla GNU/Emacs 27.1, from Debian package `emacs-lucid_27.1+1-3.1_amd64.deb`, with the following Emacs packages installed from `melpa-stable`:
```
  flycheck           31            installed             On-the-fly syntax checking
  flycheck-clj-kondo 0.0.3         installed             Add clj-kondo linter to flycheck
```

although the issue is independent of the editor used.

**Problem**

Issue #1484 (pull request #1569) implemented `:protocol-impls` analysis for `defrecord` and `deftype`. And later #1613 (pull request #1616) extended the implementation to `extend-protocol`, `extend-type`, `reify` and `specify!`.

But #1613/#1616 have two issues:

- A minor one: the protocol implementation analysis for `extend-protocol`, `extend-type`, `reify` and `specify!` are reported as being `:defined-by "extened-type"`. `defrecord` and `deftype` on the other hand are reported as `:defined-by clojure.core/defrecord` and `:defined-by clojure.core/deftype` respectively (when analyzed as CLJ, and the corresponding ClojureScript versions when analyzed as CLJS). So we lose information about what method was used to implement the protocol.
- A major one: the protocol implementation analysis for `extend-protocol` reports the wrong protocols. It reports the records and types where the protocol is being extended as the protocols implemented.

**Repro**

Execute the following commands:

```bash
$ git clone https://github.com/clj-kondo/clj-kondo.git clj-kondo.git
$ cd clj-kondo.git
$ lein uberjar
$ cat > repro.clj <<-EOF
(ns repro)

(defprotocol AProtocol
  (a-method [x y z]))

(defrecord ARecord [x])

(deftype AType [x])

(extend-protocol AProtocol
  ARecord
  (a-method
    [_x _y _z]
    nil)

  AType
  (a-method
    [_x _y _z]
    nil))
EOF
$ java -jar target/clj-kondo-2022.03.10-SNAPSHOT-standalone.jar -- --lang clj --lint repro.clj --cache false --debug --config '{:output {:format :edn :analysis {:protocol-impls true}}}'
```

clj-kondo reports the following `:protocol-impls`:

```edn
{
 ...
 :protocol-impls [{:impl-ns repro,
                   :end-row 14,
                   :name-end-col 12,
                   :protocol-ns repro,
                   :name-end-row 12,
                   :method-name a-method,
                   :name-row 12,
                   :defined-by "extend-type",
                   :protocol-name ARecord,
                   :filename "repro.clj",
                   :col 3,
                   :name-col 4,
                   :end-col 9,
                   :row 12}
                  {:impl-ns repro,
                   :end-row 19,
                   :name-end-col 12,
                   :protocol-ns repro,
                   :name-end-row 17,
                   :method-name a-method,
                   :name-row 17,
                   :defined-by "extend-type",
                   :protocol-name AType,
                   :filename "repro.clj",
                   :col 3,
                   :name-col 4,
                   :end-col 9,
                   :row 17}]}
```

As can be seen, the first entry has a `:protocol-name` value with the name of the record being used to extend the protocol implementation (`ARecord`), and the second entry the name of the type (`AType`).

Of course, the minor issue is also seen in the above output (both entries are reported as `:defined-by "extend-type"`)

**Config**

No `config.edn` file is needed to reproduce the issue.

**Expected behavior**

The expected output for `:protocol-impls` in this case would be:

```edn
{
 ...
 :protocol-impls [{:impl-ns repro,
                   :end-row 14,
                   :name-end-col 12,
                   :protocol-ns repro,
                   :name-end-row 12,
                   :method-name a-method,
                   :name-row 12,
                   :defined-by clojure.core/extend-protocol,
                   :protocol-name AProtocol,
                   :filename "repro.clj",
                   :col 3,
                   :name-col 4,
                   :end-col 9,
                   :row 12}
                  {:impl-ns repro,
                   :end-row 19,
                   :name-end-col 12,
                   :protocol-ns repro,
                   :name-end-row 17,
                   :method-name a-method,
                   :name-row 17,
                   :defined-by clojure.core/extend-protocol,
                   :protocol-name AProtocol,
                   :filename "repro.clj",
                   :col 3,
                   :name-col 4,
                   :end-col 9,
                   :row 17}]}  
```

where the name of the protocol being implemented (extended) is the right one (`AProtocol`), and the name of the method used to implement the protocol is the rigtht one (`clojure.core/extend-protocol`)

I already have a patch that fixes both issues. It also adds additional unit tests covering all the protocol implementations supported by clj-kondo as of today. I will open a PR with the changes shortly.

**Comment by borkdude** [comment]
Thank you!
